Editorial (#656)
mlswg/mls-protocol: draft-ietf-mls-protocol.md

  // first two bits of the first byte.
  v = data.next_byte()
  prefix = v >> 6
  if prefix == 3:
    raise Exception('invalid variable length integer prefix')

  length = 1 << prefix

  // Once the length is known, remove these bits and read any
  // remaining bytes.
  v = v & 0x3f
  repeat length-1 times:
    v = (v << 8) + data.next_byte()

  // Check that the encoder used the minimum bits required

>   if length > 1 && v < (1 << ((8*(length-1))-2)):

Reviewer: The fix is still buggy, here is the correct one:
```suggestion
  if prefix >= 1 && v < (1 << (8*(1 << (prefix-1))-2)):
```

Author: What's the bug in the current code?

Reviewer: In your code, all numbers between `2^14` and `2^22` will be rejected: they are represented with `prefix = 2` (hence `length = 4`). In this case, the buggy check is `v < (1 << (8*3-2))`, the fixed check is `v < (1 << (8*2-2))`.

Author: Ah, thank you!

  subsets of the group.  Each epoch has a distinct ratchet tree. It seeds the
  _key schedule_.
* A _key schedule_ that describes the chain of key derivations used to progress from

>   epoch to epoch (mainly using the _init_secret_ and _epoch_secret_) and which derives

Reviewer 2: ```suggestion
  epoch to epoch (mainly using the _init_secret_ and _epoch_secret_), as well as the derivation of
```

  a variety of other secrets (see {{epoch-derived-secrets}}) used during the current
  epoch. One of these (the _encryption_secret_) is the root of the secret tree.
* A _secret tree_ derived from the key schedule that represents shared secrets
  used by the members of the group for encrypting and authenticating messages.
  Each epoch has a distinct secret tree.

Each member of the group maintains a partial view of these components of the group's
state.  MLS messages are used to initialize these views and keep them in sync as
the group transitions between epochs.

Each new epoch is initiated with a Commit message.  The Commit instructs
existing members of the group to update their views of the ratchet tree by applying
a set of Proposals, and uses the updated ratchet tree to distribute fresh
entropy to the group.  This fresh entropy is provided only to members in the new

> epoch and not to members who have been removed, which maintains the confidentiality
> of the epoch secret (in other words, the protocol provides post-compromise security with
> respect to those members).

Reviewer 2: ```suggestion
epoch and not to members who have been removed. Commits thus maintains the property that the
the epoch secret is confidential to the members in the current epoch.
```

respect to those members).

For each Commit that adds member(s) to the group, there is a single corresponding
For each Commit that adds one or more members to the group, there is a single corresponding
Welcome message.  The Welcome message provides all the new members with the information
they need to initialize their views of the key schedule and ratchet tree, so
that these views are equivalent to the views held by other members of the group
that these views align with the views held by other members of the group
in this epoch.

In addition to defining how one epoch secret leads to the next, the key schedule
also defines a collection of secrets that are derived from the epoch secret.
For example:

* An _encryption secret_ that is used to initialize the secret tree for the
  epoch.

* A _confirmation key_ that is used to confirm that all members agree on the
  shared state of the group.

* A _resumption secret_ that members can use to prove their membership in the
  group, e.g., in the case of branching a subgroup.

Finally, an _init secret_ is derived that is used to initialize the next epoch.

> 

Reviewer 2: These seem like useful examples?  Since they show how the key schedule actually gets put to use.  Might drop `init_secret` and replace `confirmation_key` with `exporter_secret` to make it more functional.

Author: Sure, but it doesn't seem to fit in this section. The rest of the text is talking about messages and how they evolve the group state.

Reviewer 2: Maybe move them up under the `_key schedule_` bullet above?  "... derive a variety of secrets, for example: <sub-bullets>"
